### PR TITLE
Add support for JabFox on linux

### PR DESCRIPTION
**Pull request by tobiasdiez**
Makes it possible to use JabFox without any manual installation on the users' side on Linux.

----

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [ ] Manually tested changed features in running JabRef
- [ ] Screenshots added in PR description (for bigger UI changes)
- [ ] Ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

### CODE REVIEW

File: buildres/linux/jabrefHost.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/python -u
```
Comment (LyzardKing) [inline buildres/linux/jabrefHost.py]: I'd set this to
`#!/usr/bin/env python3 -u`
(do we need the -u? it sets the output to be unbuffered I think)
So it's properly set to python3
and not hardcoded to a path (https://docs.python.org/3/using/unix.html)

File: buildres/linux/jabrefHost.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+# Note that running python with the `-u` flag is required on Windows,
+# in order to ensure that stdin and stdout are opened in binary, rather
+# than text, mode.
+
+import json
+import sys
+import struct
+import subprocess
+import shlex
+import os
```
Comment (LyzardKing) [inline buildres/linux/jabrefHost.py]:
```suggestion

```

File: buildres/linux/jabrefHost.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+from pathlib import Path
+
+# We assume that this python script is located in "jabref/lib" while the executable is "jabref/bin/JabRef"
+script_dir = os.path.dirname(os.path.realpath('__file__'))
+JABREF_PATH = os.path.join(script_dir, '../bin/JabRef')
+
+logging_dir = str(Path.home()) + "/.mozilla/native-messaging-hosts/"
+if not os.path.exists(logging_dir):
+    os.makedirs(logging_dir)
+logging.basicConfig(filename=logging_dir + 'jabref_browser_extension.log')
```
Comment (LyzardKing) [inline buildres/linux/jabrefHost.py]: Use pathlib everywhere for consistency.
Change one path to point to the correct bin
```
script_dir = Path(__file__).resolve().parent.parent
JABREF_PATH = script_dir / "bin/JabRef"
logging_dir = Path.home() / ".mozilla/native-messaging-hosts/"
if not logging_dir.exists():
    logging_dir.mkdir(parents=True)
logging.basicConfig(filename=logging_dir / 'jabref_browser_extension.log')
```

File: buildres/linux/jabrefHost.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+# Send an encoded message to stdout.
+def send_message(message):
+    encoded_message = encode_message(message)
+    sys.stdout.buffer.write(encoded_message['length'])
+    sys.stdout.buffer.write(encoded_message['content'])
+    sys.stdout.buffer.flush()
+
+def add_jabref_entry(data):
+    cmd = JABREF_PATH + " -importBibtex " + "\"" + data + "\""
```
Comment (LyzardKing) [inline buildres/linux/jabrefHost.py]: Port to string from pathlib
```suggestion
    cmd = str(JABREF_PATH) + " -importBibtex " + "\"" + data + "\""
```

File: buildres/linux/jabrefHost.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+logging.info("Starting JabRef backend")
+
+try:
+    message = get_message()
+except Exception as e:
+    message = str(e)
+logging.info(str(message))
+
+if 'status' in message and message["status"] == "validate":
+    cmd = JABREF_PATH + " -version"
```
Comment (LyzardKing) [inline buildres/linux/jabrefHost.py]: Port to string from pathlib
```suggestion
    cmd = str(JABREF_PATH) + " -version"
```

File: buildres/linux/jabrefHost.py (hunk 1 above)
Comment (LyzardKing) [reply buildres/linux/jabrefHost.py]: I'd still change this line, to be sure that we use python3

File: buildres/linux/jabrefHost.py (hunk 3 above)
Comment (LyzardKing) [inline buildres/linux/jabrefHost.py]: Here I meant to say that the lines 17-23 are replaced by the ones in the comment
It's just a pathlib replacement, but github doesn't have a multiline suggestion yet

Comment (LyzardKing) [reply buildres/linux/jabrefHost.py]: If  you want I can make the changes directly to this branch..

**Comment by tobiasdiez** [comment]
Thanks @LyzardKing for the useful feedback. I've now accepted your proposed changes and also implemented your suggestion using `shutil.which`.

I take this as a positive review and merge now.

**Comment by tobiasdiez** [comment]
Yeah, the json is essentially a copy of the windows one. Probably, one can use gradle to create a corresponding file for each OS during the build, but since it was only a bit of code, I didn't bothered.

Is it possible to install the json file also during the snap installation?

**Comment by LyzardKing** [comment]
I wondered that...but the json file would have to go in /home/$USER/.mozilla/native-messaging-host/
And being a hidden folder it's not available to the snap.
It could be possible to add it as a personal file, a special hidden file available to the snap (the specific json file being only the jabref one) but it would need to be manually connected via 
`snap connect jabref:<name of the interface>`
so something like:
`snap conenct jabref:jabfox-integration`
But it all depends if the snapcraft people are willing to assign jabref a file under the mozilla directory.
I can ask on the forum

**Pull request merged**